tests/hlsl-intrinsic/byte-address-buffer/byte-address-16bit.slang

```diff
@@ -0,0 +1,51 @@
+// byte-address-16bit-load.slang
+
+// Test that loading values using 16-bit types from
+// byte-address buffers works as expected, on targets
+// that support it.
+
+//TEST(compute):COMPARE_COMPUTE_EX:-slang -compute -cpu
+//TEST(compute):COMPARE_COMPUTE_EX:-slang -compute -dx12 -use-dxil -profile cs_6_2 -render-features half
+//TEST(compute, vulkan):COMPARE_COMPUTE_EX:-slang -vk -compute -profile cs_6_2 -render-features half
+//TEST(compute):COMPARE_COMPUTE_EX:-slang -cuda -compute
+
+// Unavailable on D3D and fxc-based targets, because fxc doesn't
+// support loading anything besides `uint` from a byte-address
+// buffer.
+//
+//TEST_DISABLED(compute):COMPARE_COMPUTE_EX:-slang -compute
+
+
+//TEST_INPUT:ubuffer(data=[0x00010002 0x00030004 0x00050006 0x00070008]):name=inputBuffer
+RWByteAddressBuffer inputBuffer;
+
+//TEST_INPUT:ubuffer(data=[0xdeadbeef 0xdeadbeef 0xdeadbeef 0xdeadbeef]):name=tmpBuffer
+RWByteAddressBuffer tmpBuffer;
+
+//TEST_INPUT:ubuffer(data=[0 0 0 0], stride=4):out,name=outputBuffer
+RWStructuredBuffer<int> outputBuffer;
+
+struct Data
+{
+	int16_t	a;
+	int16_t b;
+}
+
+int test(int val)
+{
+	int offsetX = val*4;
+	int offsetY = offsetX & 0xFF;
```

Comment:
I probably should have added a comment about this.

The idea is that to a human it is clear that `offsetX` and `offsetY` will always be equal, so that the loads and stores that follow all touch the same memory, but to a downstream compiler that *won't* be clear, so it should be impossible for a compiler to optimize away a sequence like:

```
buffer[offsetX] = valueA;
valueB = buffer[offsetY];
```

and eliminate the load so you just have:

```
buffer[offsetX] = valueA;
valueB = valueA;
```

The point here is to ensure that all the operations we wrote in the function actually get executed.

In some similar codebases I've worked with in the past, tests would often have a constant buffer of inputs with names like `zero`, `one`, `two`, etc. that are always bound to the integer values the names imply. That means that you could write a test that uses `someVal + zero` and be sure the compiler couldn't optimize based on knowing that is equal to `someVal`.